### ISSUE TITLE
SGPR: specify number of gradient descent iterations between q-updates

### ISSUE THREAD

**Issue by martinsch** [GPflow/GPflow#199]
Hi all,

Sorry, I couldn't find a mailing list, otherwise I would have asked this question there rather than filing an issue. 
Is it possible in SGPR to specify how many GP hyperparameter optimization steps (ELBO gradient descent iterations) are performed before updating the q-distribution? As far as I understand the code, the q-distribution is updated in every gradient descent step, which might be slower and lead to worse local optima in some cases than updating the hyperparameters for say 10 steps and then updating the q-distribution, right?

Thanks for your help!
Martin

**Comment by markvdw** [comment]
I think that currently this should be implemented by manually fixing certain parameters, followed by optimisation.

E.g.:
`
model.kernel.lengthscales.fixed = True
model.kernel.variance.fixed = True
model.optimize()
model.kernel.variance.fixed = False
# ...etc
`

**Comment by alexggmatthews** [comment]
@martinsch Thanks for your interest in GPflow. Don't worry about posting an issue -that is how we work.

**Comment by jameshensman** [comment]
Hi @martinsch 

spgr does not 'update' the q distribution, so much as 'collapse' it. 

Picture a 2-d optimisation surface, where the horizontal axis represents the parameters (lengthscales, etc) and the vertical axis represents the parameters of the q-distribution. For any parameter setting, there's an optimal q-distribution: so for any point on the horizontal axis, we can solve the problem in a vertical slice, analytically. 

So there's a ridge, made by joining together the solutions for each point on the horizontal axis. We have an analytical expression for that ridge exactly, which is the sgpr method. I suspect that this is actually an easier optimization problem than the more general one. I wrote about this kind of problem in NIPS a few years ago [1], for conjugate -exponential families.

You can of course do exactly as @markvdw suggests using the svgp class, whose optimisation problem is exactly the two-D surface I described, before collapsing to the ridge. If you switch between fixing/unfixing parameters regularly though, the performance will be very low because the model will get recompiled each time (!). 

So I would suggest two identical models, with different fixes.

```python
m1 = GPflow.svgp.SVGP(X, Y, kern, ...)
m2 = GPflow.svgp.SVGP(X, Y, kern, ...)

m1.kern.fixed = True
m1.Z.fixed = True
m2.q_mu.fixed = True
m2.q_sqrt.fixed = True

for i in range(num_iters):
    m1.optimize(your_options_here)
    m2.set_parameter_dict(m1.get_parameter_dict())
    m2.optimize(your_options_here)
    m1.set_parameter_dict(m2.get_parameter_dict())
```

A bit messy, I admit, but unavoidable with GPflow's design. Unless we memoized tf graphs, which I'm not really ready for.  

Note that `(g|s)et_parameter_dict()` copies over all the parameters, fixed or not. 

There's nothing to stop you using stochastic optimisation for either model. I note that you could also implement natural gradients [2] for variational updates (`m1` in the above), which is something that we've kind of abandoned for the simplicity of gradient descent, even though it can be very performant in practise. 

I hope you found this ramble useful! I'd be intrigued to know the results of anything you come up with. 

[1]Hensman, Rattray and Lawrence 2012. Fast variational inference in the conjugate exponential Family. Proceedings of NIPS 2012. 
[2] Hensman, Fusi and Lawrence 2013. Gaussian processes for Big Data. Proceedings of UAI 2013.

**Comment by markvdw** [comment]
@jameshensman Can we use stochastic gradients for `SGPR`? If we subsample the data, we'd get a biased estimator of the bound and its gradients, right?

I mean, I guess this won't stop you from doing minibatches, but it's not "valid" in the sense that it converges to the right optimum...

**Comment by jameshensman** [comment]
Correct. But you can do minibatching for SVGP...

**Comment by martinsch** [comment]
> For any parameter setting, there's an optimal q-distribution: so for any point on the horizontal axis, we can solve the problem in a vertical slice, analytically. 
>  I suspect that this is actually an easier optimization problem than the more general one. 

This explanation sounds totally reasonable. Thanks!

Also the workaround with optimizing two models, keeping different sets of variables fix, sounds like a good idea. This should also work if the variational parameters are functions of each other, right? This might occur if we deal with mixtures of GPs, for instance, where the variational distribution for the mixture weights includes the parameters of the variational distribution for the inducing points, and vice versa. Or do you see a better option here?

**Comment by jameshensman** [comment]
For mixtures of GPs, you could use GPclust? 

github.com/jameshensman/GPclust

or the GPflow edition, work in progress, 
https://github.com/mathDR/GPclust/tree/gpflowbase

At the moment it doesn't do Sparse GPs, but it can do quite big numbers of time series -- 20k is no problem. We're interested in that though, if you'd like to contribute it...
